File: Silver/18870_Coordinate_Compression/Answer18870.java
import java.util.*;
import java.util.stream.*;

public class Answer18870 {
  public static void main(String[] args) {
    var s = new Scanner(System.in);
    int n = s.nextInt();
    var numbers = new ArrayList<Integer>();
    for (int i = 0; i < n; i++) {
      numbers.add(s.nextInt());
    }
    var ranks = new HashMap<Integer, Integer>();
    var numberSet = new HashSet<>(numbers);
    var numberSetList = new ArrayList<>(numberSet);
    Collections.sort(numberSetList);
    for(int i = 0; i < numberSetList.size(); i++){
      int it = numberSetList.get(i);
      ranks.put(it, i);
    }
    var result = new StringBuilder();
    for(int it : numbers){
      result.append(ranks.get(it));
      result.append(" ");
    }
    System.out.println(result);
  }
}

File: Silver/18870_Coordinate_Compression/Answer18870.rs
use std::io::stdin;
use std::collections::HashMap;
use std::collections::HashSet;

fn main(){
  stdin().read_line(&mut String::new()).unwrap();
  let mut s = String::new();
  stdin().read_line(&mut s).unwrap();
  let numbers:Vec<i32> = s.split_whitespace().map(|x| x.parse().unwrap()).collect();
  let mut ranks:HashMap<i32, usize> = HashMap::new();
  let number_set:HashSet<i32> = numbers.iter().cloned().collect();
  let mut number_set_list: Vec<i32> = number_set.into_iter().collect();
  number_set_list.sort();
  for i in 0 .. number_set_list.len() {
    let it = &number_set_list[i];
    ranks.insert(*it, i);
  }
  let mut result = String::new();
  for it in &numbers {
    result.push_str(&ranks[it].to_string());
    result.push(' ');
  }
  print!("{}", result);
}